src/neo-vm/ErrorHandle.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using System.Collections.Generic;
+namespace Neo.VM
+{
+    public class FaultState
+    {
+        public bool Rethrow;
+        public Exception Error;
+    }
+
+    public sealed class ErrorHandle
+    {
+        Stack<TryContext> tryStack = new Stack<TryContext>();
+
+        public TryContext CurContext
+        {
+            get
+            {
+                return tryStack.Peek();
+            }
+        }
+
+        public void Push(TryContext content)
+        {
+            tryStack.Push(content);
+        }
+
+        public TryContext Pop()
+        {
+            if (tryStack.Count == 0)
+                return null;
+
+            return tryStack.Pop();
+        }
+
+        public bool HandleError(ExecutionEngine engine)
+        {
+            while (tryStack.TryPeek(out TryContext context))
+            {
+                switch (context.State)
+                {
+                    case TryState.Try:
+                        {
+                            if (context.HasCatch)
+                            {
+                                context.State = TryState.Catch;
+                                ResumeContext(engine, context);
+                                engine.CurrentContext.InstructionPointer = context.CatchPointer;
+                                return true;
+                            }
+                            else
+                            {
+                                ResumeContext(engine, context);
+                                engine.ExecuteEndTryCatch(TryState.Try);
+                                engine.FaultState.Rethrow = true;
```

Comment:
@lightszero Plz, have a review. I have moved the logic of `ExecuteThrow` here.

Reply:
checked,everything is fine.

but try catch is too slow for this,maybe we should use a tag instead Exception

```c#
//slow version
                try
                {
                    Instruction instruction = CurrentContext.CurrentInstruction;
                    if (!PreExecuteInstruction() || !ExecuteInstruction() || !PostExecuteInstruction(instruction))
                    {
                        State = VMState.FAULT;
                        FaultState.Error = new InvalidOperationException("OPCode Fault:" + instruction.OpCode.ToString());
                    }
                }
                catch (CatcheableError e)
                {
                    FaultState.Error = e;
                    if (!HandleError())
                        State = VMState.FAULT;
                }
                catch
                {
                    State = VMState.FAULT;
                    FaultState.Error = new InvalidProgramException("Can't catch internal error:");
                }
```

```c#  quickVersion
                try
                {
                    Instruction instruction = CurrentContext.CurrentInstruction;
                    if (!PreExecuteInstruction() || !ExecuteInstruction() || !PostExecuteInstruction(instruction))
                    {
                        State = VMState.FAULT;
                        FaultState.Error = new InvalidOperationException("OPCode Fault:" + instruction.OpCode.ToString());
                    }
                    if(State==VMState.FAULT && FaultState.IsCatcheableError)
                    {
                          HandleError();
                    }
                }
                catch
                {
                    State = VMState.FAULT;
                    FaultState.Error = new InvalidProgramException("Can't catch internal error:");
                }


```

Reply:
Yes, It will be faster with tag

Reply:
Done

Reply:
Should we change to use `try-catch` to handle here, as some syscall may throw catchable exceptions?

Reply:
If a SYSCALL could throw catchable exceptions, how to charge for it? The resource consumption of exception handling is much higher than some SYSCALLs. That's why I asked whether `THROW` was the only place where a `CatchableException` was thrown.